### ISSUE TITLE
[display] completion in build macros is a mess

### ISSUE THREAD

**Issue by Gama11** [HaxeFoundation/haxe#8066]
There seem to be quite a few issues with completion in build macros:

```haxe
// #if macro
import haxe.macro.Context;
import haxe.macro.Expr.Field;

class Macro {
	static function build():Array<Field> {
		var fields = Context.getBuildFields();
		fields.|
		return fields;
	}
}
// #end
```

When attempting completion here, Haxe returns no results:

![](https://i.imgur.com/Z52TIaW.png)

Hm, ok, maybe it doesn't know about `getBuildFields()` because we're not in macro context, let's try `#if macro`:

![](https://i.imgur.com/2sDU1Yu.png)

Wat, now I get keywords...

Well, ok, let's get rid of `#if macro` and try not to rely on type inference by adding `:Array<Field>`... Huh, that actually works! But somehow diagnostics now isn't happy anymore (this is actually why I tried `#if macro` in the first place). :(

![](https://i.imgur.com/jcnng0Q.png)

Oh, and that diagnostic also doesn't seem to show up consistently, _sometimes_ I have to switch to another tab like `Main.hx` and back for it to show up...

At least the second issue with keyword completion showing up with `#if macro` is a regression compared to preview 4.

**Comment by Simn** [comment]
The initial example works for me.

The `#if macro` version however does not.

**Comment by Gama11** [comment]
>The initial example works for me.

Hm.. did you happen to have "interp" selected? It only seems to work on eval and neko for me... :/

**Comment by Simn** [comment]
Looking at this again, I don't understand how you expect this to work because `build` isn't a `macro` function. The compiler has no way of knowing that it's supposed to be in macro mode here.

**Comment by Gama11** [comment]
That whole situation is still quite unclear to me, I was always under the impression it's only needed for expression macros, and I often omit it from build macros. So the `macro` keyword is completely optional in build macros - except for display? Should there maybe be a compiler error or warning for omitting it?

I don't think it's just me either. If you look at https://code.haxe.org/tag/build-macro.html for instance, roughly half of the build macro entry points don't seem to use `macro`.

**Comment by RealyUniqueName** [comment]
I also find it confusing that despite being a build _macro_ it does not require a `macro` keyword.

**Comment by Simn** [comment]
I don't think this is about display in particular. Doesn't this also fail during normal compilation if you don't fence it in `#if macro`? Because otherwise `Context` has no fields and everything should fail.

**Comment by Gama11** [comment]
Works just fine in my experience, which is why everybody keeps doing it.

**Comment by Gama11** [comment]
It looks like there's indeed something wrong with fencing the whole module with `#if macro`. This doesn't work:

```haxe
#if macro
import haxe.macro.Context;
import haxe.macro.Expr.Field;

class Macro {
	static function build():Array<Field> {
		var fields = Context.getBuildFields();
		fields.
		return fields;
	}
}
#end
```

While this does:

```haxe
#if macro
import haxe.macro.Context;
import haxe.macro.Expr.Field;
#end

class Macro {
	#if macro
	static function build():Array<Field> {
		var fields = Context.getBuildFields();
		fields.
		return fields;
	}
	#end
}
```

So it seems Haxe needs to know about the existence of the `Main` type _outside of macro context_ for things to work? This might actually be the same reason why `.macro.hx` doesn't really work for build macros.

**Comment by Gama11** [comment]
Related: https://github.com/HaxeFoundation/haxe/issues/8159

**Comment by Gama11** [comment]
So #8159 works now, which is nice, but completion still doesn't work in `macro.hx` files:

![](https://i.imgur.com/ypEJB09.png)

**Comment by Gama11** [comment]
Actually, it looks like some of @Simn's recent changes must have fixed this case too. \o/

![](https://i.imgur.com/eWYSyCb.png)

It even works with no conditionals at all. So I guess all that's left to do is adding some tests.
